mpf/core/text_ui.py

```diff
@@ -144,8 +144,13 @@ def _init(self, **kwargs):
             self.machine.events.add_handler("mode_{}_started".format(mode.name), self._mode_change)
             self.machine.events.add_handler("mode_{}_stopped".format(mode.name), self._mode_change)
 
-        for player_var in self.machine.config['player_vars']:
-            self.machine.events.add_handler("player_{}".format(player_var), self._update_player)
+        if 'player_vars' in self.machine.config:
+            for player_var in self.machine.config['player_vars']:
+                self.machine.events.add_handler("player_{}".format(player_var), self._update_player)
+        else:
```

Comment:
Why is this else? Those three will always be displayed. I am not sure if the config reference above is correct at all. There is a separate setting to configure what text ui displays. I guess that should be used here.

Reply:
I do not fully understand the rational here. The old logic looked fine to me for score, ball and player number.

Player vars is a config setting to configure initial values for player vars. How does that relate to text ui? I think we got a setting somewhere to change which player vars are displayed. If that is configured I would expect it to be used here. Otherwise we should subscribe all player vars instead (as they could all be shown). Theoretically, this could also happen when we render text ui as we know which ones are displayed (obviously we would only do this once they change which should be super rare). I guess this would even speed up text ui as we would update less often.

Reply:
The link is that the only subscriptions to call the update method for the text_ui for players was the four explicitly stated: ball_ended, player_number, player_ball, and player_score.  If a random player_var changes value, e.g., aliens, it will not update on the text_ui until one of those four listed above occurs.  So if it updated aliens, but not score, it won't updated on the text_ui.

The previous PR that I put through meant that if you describe it in config as a player_var, then it will have a handler to automatically update.  Somehow, and I honestly don't know, through that magic, it makes score, player, ball all update as well.  I had them and realized I could remove them from the for loop, and they were still registered and updated correctly.  If you don't describe it in the player_vars section, then the text_ui will not update it unless one of the four mentioned above are updated, and then all update.

The best way would be to add a handler to determine did a player_XX get posted, and then that calls the _player_update method.  If there is a way to do that, then we should rip out this and my last PR on the topic and replace it.  Then we don't need to explicitly register score, ball, etc. but can just say if player_{anything} is posted in the events run _player_update.  Or as you describe at the end, create a list of what is actively being displayed and register the handlers for only what is shown.

Reply:
@jabdoa2 saw a help come through related to this. Do you have thoughts on this? If not I'll change the commit back to previous so it doesn't break. 

Doing so will mean that player_vars in text_ui won't update until score change or ball end. But text_ui being slow is better than it being broken.

Reply:
I kind of disagree. Especially on Windows or via SSH this makes MPF almost unusable. Your change will break number, score and ball (unless you add them to `player_vars`). At least the last thing needs fixing. Also `player_vars` != what is displayed. I am not sure if this really fixes the issue.

Reply:
I will just update this pr to revert to what it was prior to the last PR on this tomorrow. That means score, player, and ball have explicit handlers and will be the only events that redraw player_vars in text UI (plus ball_end).

This code, and as I said, I don't know how updates score, player, and ball without an explicit handler, but it does as that is what's in dev branch right now. But it requires player_vars section in config to make that magic happen.

Ultimately this needs a way to add a wildcard handler for player_* to register all to call the update. But I have tried and can't get a dynamic runtime replacement. Or some alternative way to register. Otherwise player_vars won't update unless those three are called.